File: app/src/main/java/org/lineageos/recorder/service/GoodQualityRecorder.java
/*
 * Copyright (C) 2021 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.recorder.service;

import android.Manifest;
import android.media.MediaRecorder;

import androidx.annotation.RequiresPermission;

import java.io.IOException;
import java.nio.file.Path;

public class GoodQualityRecorder implements SoundRecording {
    private static final String FILE_NAME_EXTENSION_AAC = "m4a";
    private static final String FILE_MIME_TYPE_AAC = "audio/mp4a-latm";

    private MediaRecorder mRecorder = null;
    private boolean mIsPaused = false;

    @Override
    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
    public void startRecording(Path path) throws IOException {
        mRecorder = new MediaRecorder();
        mRecorder.setOutputFile(path.toFile());
        mRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        mRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
        mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);

        mRecorder.prepare();
        mRecorder.start();
    }

    @Override
    public boolean stopRecording() {
        if (mRecorder == null) {
            return false;
        }

        if (mIsPaused) {
            mIsPaused = false;
            mRecorder.resume();
        }

        // needed to prevent app crash when starting and stopping too fast
        try {
            mRecorder.stop();
        } catch (RuntimeException rte) {
            return false;
        } finally {
            mRecorder.release();
        }
        return true;
    }

    @Override
    public boolean pauseRecording() {
        if (mIsPaused || mRecorder == null) {
            return false;
        }

        mIsPaused = true;
        mRecorder.pause();
        return true;
    }

    @Override
    public boolean resumeRecording() {
        if (!mIsPaused || mRecorder == null) {
            return false;
        }

        mRecorder.resume();
        mIsPaused = false;
        return true;
    }

    @Override
    public int getCurrentAmplitude() {
        return mRecorder == null ? 0 : mRecorder.getMaxAmplitude();
    }

    @Override
    public String getMimeType() {
        return FILE_MIME_TYPE_AAC;
    }

    @Override
    public String getFileExtension() {
        return FILE_NAME_EXTENSION_AAC;
    }
}

File: app/src/main/java/org/lineageos/recorder/service/GoodQualityRecorder.kt
/*
 * SPDX-FileCopyrightText: 2021-2024 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.recorder.service

import android.Manifest.permission
import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import androidx.annotation.RequiresPermission
import java.io.IOException
import java.nio.file.Path

class GoodQualityRecorder(private val context: Context) : SoundRecording {
    private var recorder: MediaRecorder? = null
    private var isPaused = false

    @RequiresPermission(permission.RECORD_AUDIO)
    @Throws(IOException::class)
    override fun startRecording(path: Path) {
        recorder = (if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            MediaRecorder(context)
        } else {
            @Suppress("Deprecation")
            MediaRecorder()
        }).apply {
            setOutputFile(path.toFile())
            setAudioSource(MediaRecorder.AudioSource.MIC)
            setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
            setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
            prepare()
            start()
        }
    }

    override fun stopRecording() = recorder?.let {
        if (isPaused) {
            isPaused = false
            it.resume()
        }

        // needed to prevent app crash when starting and stopping too fast
        try {
            it.stop()
        } catch (rte: RuntimeException) {
            return false
        } finally {
            it.release()
        }

        return true
    } ?: false

    override fun pauseRecording() = recorder?.let {
        if (isPaused) {
            return false
        }

        isPaused = true

        it.pause()

        true
    } ?: false

    override fun resumeRecording() = recorder?.let {
        if (!isPaused) {
            return false
        }

        it.resume()

        isPaused = false

        true
    } ?: false

    override val currentAmplitude: Int
        get() = recorder?.maxAmplitude ?: 0

    override val fileExtension = "m4a"

    override val mimeType = "audio/mp4a"
}
